### PR TITLE
added dust tests and lenderActions changes

### PR DESCRIPTION
**Pull request by ith-harvey**
## Description
* Created `testRemoveQuoteDust` and `testMoveQuoteDust` tests
* added dust limit checks in the following functions (ensuring that the bucket) is not left with deposit after the action):
  * `removeMaxDeposit()`
  * `addQuoteToken()`
  * `moveQuoteToken()`
 
## Purpose
* Resolves `Kirill L-01` as detailed here:
```
## [L-01] Dust checks for quote tokens are inconsistent

`LenderActions::RemoveDepositParams::dustLimit` is defined but is left unused in `moveQuoteToken`
and `removeQuoteToken`.

Also, `LenderActions::moveQuoteToken` constrains `maxAmountToMove` from below with `quoteTokenScale` as a dust limit.
However, `LenderActions::removeQuoteToken` is missing such a constraint.

### Recommendation

1. Remove `LenderActions::RemoveDepositParams::dustLimit`.
2. Add dust limit check to the `LenderActions::removeQuoteToken`.

if (params_.maxAmount < poolState_.quoteTokenScale)
    revert DustAmountNotExceeded();


3. Simplify constraint in `LenderActions::moveQuoteToken`.

if (params_.maxAmountToMove == 0)
    revert InvalidAmount();
-if (params_.maxAmountToMove != 0 && params_.maxAmountToMove < poolState_.quoteTokenScale)
+if (params_.maxAmountToMove < poolState_.quoteTokenScale)
    revert DustAmountNotExceeded();
```
## Impact

old
```
addQuoteToken                        | 1950            | 252735 | 169787 | 723088 | 944     |
moveQuoteToken                       | 1918            | 149390 | 107324 | 634633 | 41      |
```

new
```
addQuoteToken                        | 1950            | 253720 | 170124 | 723195 | 949     |
moveQuoteToken                       | 1918            | 146633 | 107386 | 634829 | 44      |
```

### size
old
```
  ERC721Pool               -  24,069B  (97.93%)
  ERC20Pool                -  23,360B  (95.05%)
  PositionManager          -  17,801B  (72.43%)
  TakerActions             -  15,866B  (64.56%)
  PositionNFTSVG           -  15,271B  (62.14%)
  PoolInfoUtils            -  14,399B  (58.59%)
  LenderActions            -  10,715B  (43.60%)
  BorrowerActions          -   9,358B  (38.08%)
  SettlerActions           -   9,349B  (38.04%)
  KickerActions            -   9,153B  (37.24%)
  PoolCommons              -   9,089B  (36.98%)
  RewardsManager           -   8,693B  (35.37%)
  PoolInfoUtilsMulticall   -   8,539B  (34.74%)
  ERC721                   -   4,276B  (17.40%)
  ERC721PoolFactory        -   3,742B  (15.23%)
  LPActions                -   2,952B  (12.01%)
  ERC20PoolFactory         -   2,564B  (10.43%)
```

new
```
  ERC721Pool               -  24,069B  (97.93%)
  ERC20Pool                -  23,360B  (95.05%)
  PositionManager          -  17,801B  (72.43%)
  TakerActions             -  15,866B  (64.56%)
  PositionNFTSVG           -  15,271B  (62.14%)
  PoolInfoUtils            -  14,399B  (58.59%)
  LenderActions            -  10,823B  (44.04%)
  BorrowerActions          -   9,358B  (38.08%)
  SettlerActions           -   9,349B  (38.04%)
  KickerActions            -   9,153B  (37.24%)
  PoolCommons              -   9,089B  (36.98%)
  RewardsManager           -   8,693B  (35.37%)
  PoolInfoUtilsMulticall   -   8,539B  (34.74%)
  ERC721                   -   4,276B  (17.40%)
  ERC721PoolFactory        -   3,742B  (15.23%)
  LPActions                -   2,952B  (12.01%)
  ERC20PoolFactory         -   2,564B  (10.43%)
  ERC20                    -   2,114B  (8.60%)
```
## Tasks

- [x] Changes to protocol contracts are covered by unit tests executed by CI.
- [x] Protocol contract size limits have not been exceeded.
- [x] Gas consumption for impacted transactions have been compared with the target branch, and nontrivial changes cited in the _Impact_ section above.
- [x] Scope labels have been assigned as appropriate.
- [x] Invariant tests have been manually executed as appropriate for the nature of the change.

### CODE REVIEW

**Review by mattcushman — APPROVED** [review]
lgtm

File: src/libraries/external/LenderActions.sol [hunk 1]
```diff
@@ -170,6 +170,8 @@ library LenderActions {
         uint256 bucketPrice           = _priceAt(params_.index);
         uint256 addedAmount           = params_.amount;
 
+        if (bucketDeposit + addedAmount < poolState_.quoteTokenScale) revert DustAmountNotExceeded();
```
Comment (EdNoepel) [inline src/libraries/external/LenderActions.sol]: I don't see the purpose of this line.  https://github.com/ajna-finance/contracts/blob/develop/src/base/Pool.sol#L164 already ensures the bucket cannot have between 0 and the token scale in an `addQuoteToken` call.

File: tests/forge/unit/ERC721Pool/ERC721PoolInputValidation.t.sol [hunk 2]
```diff
@@ -28,8 +28,8 @@ contract ERC721PoolBorrowTest is ERC721HelperContract {
     }
 
     function testValidateMoveQuoteTokenInput() external tearDown {
-        // revert on zero amount
-        vm.expectRevert(IPoolErrors.InvalidAmount.selector);
+        // revert on innsufficientLiquidity amount if none in bucket
```
Comment (EdNoepel) [inline tests/forge/unit/ERC721Pool/ERC721PoolInputValidation.t.sol]: nit: spelling _insufficient_

File: src/libraries/external/LenderActions.sol [hunk 3]
```diff
@@ -311,6 +311,9 @@ library LenderActions {
         // revert if (due to rounding) the awarded LP in to bucket is 0
         if (toBucketLP_ == 0) revert InsufficientLP();
 
+        // revert if toBucket deposit is below dust limit
+        if (vars.toBucketDeposit + movedAmount_ < poolState_.quoteTokenScale) revert DustAmountNotExceeded();
```
Comment (EdNoepel) [inline src/libraries/external/LenderActions.sol]: This seems superfluous.  Line 252 prevents moving less than than token precision/dust amount.

File: tests/forge/interactions/ERC20DustCheck.t.sol [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+import { ERC20Pool }        from 'src/ERC20Pool.sol';
+import { ERC20PoolFactory } from 'src/ERC20PoolFactory.sol';
+import { IPoolErrors }      from 'src/interfaces/pool/commons/IPoolErrors.sol';
+
+import 'src/PoolInfoUtils.sol';
+
+import "./BalancerUniswapExample.sol";
+import "./UniswapTakeExample.sol";
+
+contract ERC20TakeWithExternalLiquidityTest is Test {
```
Comment (EdNoepel) [inline tests/forge/interactions/ERC20DustCheck.t.sol]: Why is contract named as such?  There is no _take_ operation.  Should likely relate to filename.

**Review by EdNoepel — CHANGES_REQUESTED** [review]
_LenderActions_ lines 173 and 315 seem redundant, and unit tests pass with them commented-out.  If they are needed, we should create a unit test to demonstrate why.

File: tests/forge/interactions/ERC20DustCheck.t.sol (hunk 4 above)
Comment (EdNoepel) [reply tests/forge/interactions/ERC20DustCheck.t.sol]: There are also some unnecessary imports and constants.

Comment (EdNoepel) [reply tests/forge/interactions/ERC20DustCheck.t.sol]: Also probably doesn't belong in the `interactions` folder, as we're not testing interactions with external liquidity.

File: src/libraries/external/LenderActions.sol (hunk 3 above)
Comment (ith-harvey) [reply src/libraries/external/LenderActions.sol]: Agree, fixed -> https://github.com/ajna-finance/contracts/pull/978/commits/bb64f2d5c74f654275a0d2b365d465f73b0d3197

File: tests/forge/unit/ERC721Pool/ERC721PoolInputValidation.t.sol (hunk 2 above)
Comment (ith-harvey) [reply tests/forge/unit/ERC721Pool/ERC721PoolInputValidation.t.sol]: Agree, wrong comment entirely -> https://github.com/ajna-finance/contracts/pull/978/commits/bb64f2d5c74f654275a0d2b365d465f73b0d3197

File: src/libraries/external/LenderActions.sol (hunk 1 above)
Comment (ith-harvey) [reply src/libraries/external/LenderActions.sol]: Agree, redundant -> https://github.com/ajna-finance/contracts/commit/bb64f2d5c74f654275a0d2b365d465f73b0d3197

**Comment by EdNoepel** [comment]
Also, merge target should be `develop` branch, not `master`.

**Comment by ith-harvey** [comment]
> Also, merge target should be `develop` branch, not `master`.

I mentioned this briefly in our standup the other day. First step was the team agreeing that this was an issue by creating a unit test off of `master`. <- Seem to have consensus there.

Now will move it to off of develop so we can get it merged there : )

File: tests/forge/interactions/ERC20DustCheck.t.sol (hunk 4 above)
Comment (ith-harvey) [reply tests/forge/interactions/ERC20DustCheck.t.sol]: Agree, thank you for your help. Resolved -> https://github.com/ajna-finance/contracts/pull/978/commits/d3c304710b62a5303ce54e78c48f3c80998f5e28

**Comment by ith-harvey** [comment]
Changes we're moved onto this branch to simply git commits -> https://github.com/ajna-finance/contracts/pull/980

**Pull request closed without merging**